clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/ConnectionSource.java

```diff
@@ -126,12 +129,12 @@ public LeasedConnection connect(Properties connectionProperties) throws Connecti
         dynamicTopologyEntity.setListener(new DynamicTopologyEntity.Listener() {
           @Override
           public void onNodeRemoval(Cluster cluster, UID stripeUID, Node removedNode) {
-            removedNode.getEndpoints().forEach(e -> servers.remove(e.getAddress()));
+            removedNode.getEndpoints().forEach(e -> servers.remove(e.getHostPort().createInetSocketAddress()));
           }
 
           @Override
           public void onNodeAddition(Cluster cluster, UID addedNodeUID) {
-            servers.add(cluster.determineEndpoint(addedNodeUID, servers).get().getAddress());
+            servers.add(cluster.determineEndpoint(addedNodeUID, servers.stream().map(HostPort::create).collect(toList())).get().getHostPort().createInetSocketAddress());
```

Comment:
I hate this code... I'm going to keep looking at it to see if I can rework the wider area in to something better. Suggestions are welcome.

Reply:
wdym ? tou hate determineEndpoint or the fact that ehcache is not using HostPort ?

Reply:
Sorry... the "I hate this code" was about the code before this change.